### PR TITLE
Photon Controller Driver: Added Pre-checks and fixed IP retrieval for machine

### PR DESCRIPTION
**Pull request by faijaz**
Added Pre-checks and fixed IP retrieval for machine.

### CODE REVIEW

**Comment by AlainRoy** [comment]
I'm curious: have you been keeping the vmware/docker-machine branch up to date with the upstream from docker? I don't know how much churn there is for you... or will be in the future.

File: drivers/photoncontroller/photoncontroller.go [hunk 1]
```diff
@@ -74,15 +72,9 @@ func (d *Driver) GetCreateFlags() []mcnflag.Flag {
 		mcnflag.StringFlag{
 			Name:   "photon-endpoint",
 			Usage:  "Photon Controller endpoint in format: http://<ip address>:<port>",
-			Value:  defaultEndpoint,
 			EnvVar: "PHOTON_ENDPOINT",
 		},
 		mcnflag.StringFlag{
-			Name:   "photon-tenant",
```
Comment (AlainRoy) [inline drivers/photoncontroller/photoncontroller.go]: Why did you remove the tenant? Do you not need it if you have the project?

File: drivers/photoncontroller/photoncontroller.go [hunk 2]
```diff
@@ -267,6 +258,30 @@ func (d *Driver) Create() error {
 }
 
 func (d *Driver) PreCreateCheck() error {
+	if d.PhotonEndpoint == "" {
+		return fmt.Errorf("Photon controller endpoint is not provided.")
```
Comment (AlainRoy) [inline drivers/photoncontroller/photoncontroller.go]: In this method, all the messages should saw "was not provided", not "is not provided".

If you want to be really cool, the error message could tell people the option to use. Something like:

"The Photon controller endpoint was not specified. Use the --photon-endpoint option to specify it". (Or something like that.

File: drivers/photoncontroller/photoncontroller.go [hunk 3]
```diff
@@ -385,3 +357,49 @@ func (d *Driver) Restart() error {
...
+	d.IPAddress = ""
+	numRetries := 0
+	for numRetries < maxRetries {
+		vmNetworksTask, err := client.VMs.GetNetworks(d.VMId)
+		if (err != nil) {
+			return fmt.Errorf("Error creating task for get VM networks: ", err)
+		}
+
+		// Waiting for get VM networks task completion
+		vmNetworksTask, err = client.Tasks.Wait(vmNetworksTask.ID)
```
Comment (AlainRoy) [inline drivers/photoncontroller/photoncontroller.go]: I don't know how our API works. Do you need to specify a timeout? Is there a default timeout?

File: drivers/photoncontroller/photoncontroller.go [hunk 4]
```diff
@@ -385,3 +357,49 @@ func (d *Driver) Restart() error {
...
+		if (err != nil) {
+			return fmt.Errorf("Get VM networks taks not completed: ", err)
```
Comment (AlainRoy) [inline drivers/photoncontroller/photoncontroller.go]: It's not that it didn't complete, but that it failed. Perhaps "Failed to get IP address for VM <id>". I think that specifying the ID may be helpful for future debugging.

File: drivers/photoncontroller/photoncontroller.go [hunk 5]
```diff
@@ -385,3 +357,49 @@ func (d *Driver) Restart() error {
...
+
+		if d.IPAddress != "" {
+			break;
+		}
+		numRetries++
+		time.Sleep(1 * time.Second)
+	}
+
+	if d.IPAddress == "" {
+		return fmt.Errorf("Fail to retrieve VM IP.")
```
Comment (AlainRoy) [inline drivers/photoncontroller/photoncontroller.go]: Another place where the error should include the id of the VM.

File: drivers/photoncontroller/photoncontroller.go [hunk 6]
```diff
@@ -385,3 +357,49 @@ func (d *Driver) Restart() error {
...
+		}
+
+		networkConnections := vmNetworksTask.ResourceProperties.(map[string]interface{})
+		networks := networkConnections["networkConnections"].([]interface{})
+
+		for _, nt := range networks {
+			network := nt.(map[string]interface{})
+			ipAddressValue, ok := network["ipAddress"]
+			if ok && ipAddressValue != nil && ipAddressValue.(string) != "" {
+				matched := regexp.MustCompile("172.*").MatchString(ipAddressValue.(string))
```
Comment (AlainRoy) [inline drivers/photoncontroller/photoncontroller.go]: Why is 172.* invalid? It's not obvious to me. When I was at F5 networks, all my VMs had IP addresses that began with 172. 

In the Kubernetes example, I had a special interface where I knew that it shouldn't be a 172 address, but in this case I don't think you know that. I'll stop by and chat about it.

File: drivers/photoncontroller/photoncontroller.go (hunk 1 above)
Comment (faijaz) [inline drivers/photoncontroller/photoncontroller.go]: No, I don't need it if I have project.

File: drivers/photoncontroller/photoncontroller.go (hunk 3 above)
Comment (faijaz) [inline drivers/photoncontroller/photoncontroller.go]: Yes, there is a default timeout. That should be sufficient for now as our tests use the same timeout value today.

File: drivers/photoncontroller/photoncontroller.go (hunk 5 above)
Comment (faijaz) [inline drivers/photoncontroller/photoncontroller.go]: VM IP is printed just above all these steps in the same call. That's why no need for IP printing here.

File: drivers/photoncontroller/photoncontroller.go (hunk 6 above)
Comment (faijaz) [inline drivers/photoncontroller/photoncontroller.go]: Fixed it with network check.

File: drivers/photoncontroller/photoncontroller.go (hunk 4 above)
Comment (faijaz) [inline drivers/photoncontroller/photoncontroller.go]: I wanted to know that the problem was with the get network task not in the IP parsing which is retrieved from task result below.

File: drivers/photoncontroller/photoncontroller.go (hunk 2 above)
Comment (faijaz) [inline drivers/photoncontroller/photoncontroller.go]: Fixed them in the other check-in.

**Comment by AlainRoy** [comment]
Looks good to me!
